host-bmc: Implement unique leaf nodes

Jino26 opened the pull request:
This commit adds support for implementing unique leaf nodes for mex & splitter dbus objects. The unique key values are generated by using a combination of chassis number and preceding entity instance numbers and then prepending them to each leaf node. This ensures that leaf nodes remain unique so that the redfish client can easily distinguish the leaf nodes and populate them correctly.

Tested on MEX & splitter systems.

Jino26 commented:
[bus_tree_pldm_mex_unique_leafnodes.txt](https://github.com/user-attachments/files/22332232/bus_tree_pldm_mex_unique_leafnodes.txt)

Pavithrab7 commented on host-bmc/utils.cpp:
```diff
@@ -56,6 +57,63 @@ Entities getParentEntites(const EntityAssociations& entityAssoc)
+    else
+    {
+        return true;
+    }
```
this else can be simplified to return std::regex_search(fullPath, chassisRegex);

Pavithrab7 commented on host-bmc/utils.cpp:
```diff
@@ -56,6 +57,63 @@ Entities getParentEntites(const EntityAssociations& entityAssoc)
+
+        if (chassisFound)
+        {
+            size_t pos = s.find_last_not_of("0123456789");
```
Add a comment here on why this extraction logic?

Jino26 replied:
Hi Pavithra, changed the logic so this is not used anymore.

Jino26 committed:
host-bmc: Implement unique leaf nodes

This commit adds support for implementing unique leaf nodes
for mex & splitter dbus objects. The unique key values are
generated by using a combination of chassis number and preceding
entity instance numbers and then prepending them to each leaf
node. This ensures that leaf nodes remain unique so that the
redfish client can easily distinguish the leaf nodes and populate
them correctly.

Tested on MEX & splitter systems.

Signed-off-by: Jino Abraham <jinoabraham26@gmail.com>

Jino26 commented:
### **Unique leaf nodes:**
└─ /xyz/openbmc_project/inventory/system/chassis15363
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_fan1
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_fan2
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_fan3
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_fan4
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1
    │     │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_connector1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_connector2
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_power_converter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot1
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot1/15363111_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot2
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot2/15363112_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot3
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot3/15363113_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot4
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot4/15363114_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot5
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot5/15363115_adapter1
    │     │   └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot6
    │     │     └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot6/15363116_adapter1
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2
    │     │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_connector1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_connector2
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_power_converter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot1
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot1/15363221_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot2
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot2/15363222_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot3
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot3/15363223_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot4
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot4/15363224_adapter1
    │     │   ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot5
    │     │   │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot5/15363225_adapter1
    │     │   └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot6
    │     │     └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot6/15363226_adapter1
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_motherboard3
    │     │ └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_motherboard3/153633_system_management_module1
    │     ├─ /xyz/openbmc_project/inventory/system/chassis15363/15363_powersupply1
    │     └─ /xyz/openbmc_project/inventory/system/chassis15363/15363_powersupply2
    ├─ /xyz/openbmc_project/led
    │ └─ /xyz/openbmc_project/led/groups
    │   ├─ /xyz/openbmc_project/led/groups/host_lamp_test
    │   └─ /xyz/openbmc_project/led/groups/system
    │     └─ /xyz/openbmc_project/led/groups/system/chassis15363
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_fan1
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_fan2
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_fan3
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_fan4
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1
    │       │ └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_connector1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_connector2
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_power_converter1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot2
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot3
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot4
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot5
    │       │   └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot1/153631_io_module1/1536311_slot6
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2
    │       │ └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_connector1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_connector2
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_power_converter1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot1
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot2
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot3
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot4
    │       │   ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot5
    │       │   └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_logical_slot2/153632_io_module2/1536322_slot6
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_motherboard3
    │       │ └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_motherboard3/153633_system_management_module1
    │       ├─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_powersupply1
    │       └─ /xyz/openbmc_project/led/groups/system/chassis15363/15363_powersupply2

Jino26 commented:
[pldm_tree_mex_unique_leafnodes_latest.txt](https://github.com/user-attachments/files/22401606/pldm_tree_mex_unique_leafnodes_latest.txt)

[bus_pldm_final_tree.txt](https://github.com/user-attachments/files/22559832/bus_pldm_final_tree.txt)

[splitter_w_unique_leaf_nodes.txt](https://github.com/user-attachments/files/22588362/splitter_w_unique_leaf_nodes.txt)

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -59,14 +59,18 @@ Entities getParentEntites(const EntityAssociations& entityAssoc)
 
     bool found = false;
+    std::string leafNode;
+    fs::path p;
```
use readable variable name.

Jino26 replied:
Done.

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -92,11 +107,24 @@ void addObjectPathEntityAssociations(
             {
                 continue;
             }
+            /*If chassis number is not empty, it means we've encountered
+              numbered chassis. So we prepend previous entity instance id and
+              chassis number to prepare unique leafnodes
+            */
```
please use consistent commenting style throughout the entire file. I see that the style was being changed down below.

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -92,11 +107,24 @@ void addObjectPathEntityAssociations(
+            if (!chassisNum.empty() &&
+                node_entity.entity_type != PLDM_ENTITY_SYSTEM_CHASSIS)
+            {
+                leafNode = (chassisNum + previousEntityInstanceNum) + "_" +
+                           entityNode;
```
prefer `std::format` to format `arbitrary strings` rather than string appends manually.

Jino26 replied:
Done.

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -123,19 +151,24 @@ void addObjectPathEntityAssociations(
+        leafNode =
+            chassisNum.empty()
+                ? entityNode
+                : ((chassisNum + previousEntityInstanceNum) + "_" + entityNode);
```
std::format here as well

Jino26 replied:
Done.

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -77,6 +81,17 @@ void addObjectPathEntityAssociations(
+    std::string entityInstanceNum =
+        std::to_string(node_entity.entity_instance_num);
+
+    /* If we get entity type as chassis, it means we encountered a mex or
+     * splitter chassis*/
```
terms like mex and splitter are IBM specific , avoid quoting those in comments since community does not understand them.

manojkiraneda commented on host-bmc/utils.cpp:
```diff
@@ -221,6 +272,14 @@ void updateEntityAssociation(
+         * will become chassis15873/15873_motherboard3"*/
+        if (!chassisNum.empty() && !paths.empty())
+        {
+            paths.front() = chassisNum + "_" + paths.front();
```
std::format

manojkiraneda reviewed (approved):
This commit needs changes if we put it to upstream , because for P12 bmc PLDM will not be hosting any cores at all (so all the checks on the chassis numbering will be going away) soon. This is just a temp solution until Inventory & Hostboot migrates to newer way of handling the cores & core count.
